src/main/java/org/sitenv/vocabularies/validation/validators/nodetypes/RequiredNodeValidator.java

```diff
@@ -0,0 +1,98 @@
+import org.sitenv.vocabularies.validation.dto.enums.VocabularyValidationResultLevel;
+import org.sitenv.vocabularies.validation.utils.ConfiguredExpressionFilter;
+import org.sitenv.vocabularies.validation.utils.XpathUtils;
+import org.sitenv.vocabularies.validation.validators.NodeValidator;
+import org.springframework.stereotype.Component;
+
+import com.ximpleware.VTDNav;
+
+@Component(value = "RequiredNodeValidator")
+public class RequiredNodeValidator extends NodeValidator {
+	
+	private static Logger LOGGER = Logger.getLogger(RequiredNodeValidator.class); 
+	private static final String AT_SIGN = "@";
+	private static final short ONE = 1;
+
+	@Override
+	public List<VocabularyValidationResult> validateNode(ConfiguredValidator configuredValidator, XPath xpath,
+			VTDNav vn, int nodeIndex, ConfiguredExpressionFilter filter, String xpathExpression) {	
+		boolean hasNode = false;
+
+    	LOGGER.debug("Entered RequiredNodeValidator.validateNode() with xpathExpression = " + xpathExpression);
+    	
+        try {
+        	String nodeName = configuredValidator.getRequiredNodeName();
+//        	LOGGER.debug("nodeName: " + nodeName);
+        	
+        	if (nodeName.startsWith(AT_SIGN)) { // It's an attribute such as @unit
+        		String attrName = nodeName.substring(ONE);
+        		hasNode = vn.hasAttr(attrName);
+//        		LOGGER.debug("Found Attribute: " + attrName);
+        	} else if (StringUtils.isNotBlank(nodeName)) { // It's an element
+        		hasNode = vn.matchElement(nodeName);
+//        		LOGGER.debug("Found Element: " + nodeName);        		
+        	} else {
+        		LOGGER.error("Found null or blank nodeName: " + nodeName); 
+        	}
+        	
+//        	LOGGER.debug("hasNode = " + hasNode);
+        } catch (Exception e) {
+			throw new RuntimeException("ERROR parsing document with given XPath expression: " + e.getMessage());
+        }
+
+		NodeValidationResult nodeValidationResult = new NodeValidationResult();
+        nodeValidationResult.setErrorOffset(vn.getTokenOffset(vn.getCurrentIndex()));
+        nodeValidationResult.setValidatedDocumentXpathExpression(xpathExpression);
+        nodeValidationResult.setRequestedNode(configuredValidator.getRequiredNodeName());
+        nodeValidationResult.setConfiguredValidationMessage(configuredValidator.getValidationMessage());
+        nodeValidationResult.setRuleID(configuredValidator.getId());
+        
+        if(hasNode) {
+        	nodeValidationResult.setValid(true);
+        }
+        
+		return buildVocabularyValidationResults(nodeValidationResult, configuredValidator.getConfiguredValidationResultSeverityLevel(), filter, vn);    
+	}
+
+	@Override
+	protected List<VocabularyValidationResult> buildVocabularyValidationResults(
+			NodeValidationResult nodeValidationResult,
+			ConfiguredValidationResultSeverityLevel configuredValidationResultSeverityLevel, ConfiguredExpressionFilter filter, VTDNav nav) {
```

Comment:
if filter object is not used, do you need it be passed ?

Reply:
Yes it needs to be passed because RequiredNodeValidator extends abstract class NodeValidator, so it has to implement NodeValidator's abstract method buildVocabularyValidationResults(...), which needs the filter object as an input parameter.

Reply:
Did not notice the "@Override".  Also , I see a change in the method signature in the latest version for NodeValidator .   The filter and nav params are removed.  Hope you noticed it.
I am not sure what version you took on but just want to let you know.

Reply:
The ETT Validator never has the last 2 params ConfiguredExpressionFilter filter, VTDNav nav in the method buildVocabularyValidationResults(...).  They are features we added for the CCDA Validator to significantly improve its performance.

Reply:
ok got it.